EventSubscriber/NavigationInvalidationSubscriber.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\HeadlessBundle\EventSubscriber;
+
+use PHPCR\SessionInterface;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\DocumentInspector;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder;
+use Sulu\Bundle\HttpCacheBundle\Cache\CacheManager;
+use Sulu\Component\DocumentManager\Event\PublishEvent;
+use Sulu\Component\DocumentManager\Event\RemoveEvent;
+use Sulu\Component\DocumentManager\Event\RemoveLocaleEvent;
+use Sulu\Component\DocumentManager\Event\UnpublishEvent;
+use Sulu\Component\DocumentManager\Events;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class NavigationInvalidationSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var CacheManager|null
+     */
+    private $cacheManager;
+
+    /**
+     * @var DocumentInspector
+     */
+    private $documentInspector;
+
+    /**
+     * @var PropertyEncoder
+     */
+    private $propertyEncoder;
+
+    /**
+     * @var SessionInterface
+     */
+    private $defaultSession;
+
+    /**
+     * @var SessionInterface
+     */
+    private $liveSession;
+
+    public function __construct(
+        ?CacheManager $cacheManager,
+        PropertyEncoder $propertyEncoder,
+        DocumentInspector $documentInspector,
+        SessionInterface $defaultSession,
+        SessionInterface $liveSession
+    ) {
+        $this->cacheManager = $cacheManager;
+        $this->propertyEncoder = $propertyEncoder;
+        $this->documentInspector = $documentInspector;
+        $this->defaultSession = $defaultSession;
+        $this->liveSession = $liveSession;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            Events::PUBLISH => ['invalidateNavigationContextBeforePublishing', 8192],
+            Events::UNPUBLISH => ['invalidateNavigationContextBeforeUnpublishing', 8192],
+            Events::REMOVE => ['invalidateNavigationContextBeforeRemoving', 8192],
+            Events::REMOVE_LOCALE => ['invalidateNavigationContextBeforeRemovingLocale', 8192],
```

Comment:
This listeners should just collect the data.

And a `Flush` subscriber with alow priority shoudl then call the CacheManager.

Do not forget to implement kernel.reset event to empty the array and empty the array.